File: Decorator.py
"""
Decorator
Attach additional responsibilities to an object dynamically.
Decorators provide a flexible alternative to subclassing for extending functionality.
"""

from abc import ABC, abstractmethod


class Component(ABC):
    @abstractmethod
    def method(self):
        pass

class Base(Component):
    def method(self):
        print("Base")

class Decorator(Component):
    def __init__(self, component):
        self.component:Component = component

    def method(self):
        self.component.method()

class ConcreteDecorator1(Decorator):
    def method(self):
        super().method()
        print("Concrete Decorator 1")

class ConcreteDecorator2(Decorator):
    def method(self):
        super().method()
        print("Concrete Decorator 2")


if __name__ == "__main__":
    b = Base()
    d1 = ConcreteDecorator1(b)
    d2 = ConcreteDecorator2(d1)

    d2.method()

File: Decorator.java
/*****
 * Decorator
 * Attach additional responsibilities to an object dynamically.
 * Decorators provide a flexible alternative to subclassing for extending functionality.
*****/


interface Component { void method(); }

class Base implements Component {
    @Override
    public void method() { System.out.println("Base"); }
}

abstract class Decorator_ implements Component {
    protected Component component;

    public Decorator_(Component component) { this.component = component; }

    @Override
    public void method() { component.method(); }
}

class ConcreteDecorator1 extends Decorator_ {
    public ConcreteDecorator1(Component component) { super(component); }

    @Override
    public void method() {
        super.method();
        System.out.println("Concrete Decorator 1");
    }
}

class ConcreteDecorator2 extends Decorator_ {
    public ConcreteDecorator2(Component component) { super(component); }

    @Override
    public void method() {
        super.method();
        System.out.println("Concrete Decorator 2");
    }
}


public class Decorator {
    public static void main(String[] args) {
        Component b = new Base();
        Component d1 = new ConcreteDecorator1(b);
        Component d2 = new ConcreteDecorator2(d1);

        d2.method();
    }
}
